source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,659 @@
+import std.algorithm.iteration : map;
+import std.range : iota;
+import std.traits;
+
+template ParameterAttribute(alias F, int i, int j)
+{
+  static if (is(typeof(F) P == __parameters)) {
+    alias ParameterAttribute = Alias!(__traits(getAttributes, P[i..i+1])[j]);
+  }
+}
+
+/**
+    Return a Function object that captures all the aspects of Fun, using the
+    value of localSymbol to represent the return and parameter types, and the
+    UDAs.
+
+    Params:
+    fun = a function
+    localSymbol = a string mixin that represents fun in the caller's context
+
+*/
+
+Function refract(alias fun, string localSymbol)()
+  if (is(typeof(fun) == function))
+{
+  Function model = {
+  name: __traits(identifier, fun),
+  localSymbol: localSymbol,
+  returnType: "std.traits.ReturnType!("~localSymbol~")",
+  parameters: refractParameterList!(fun, localSymbol),
+  udas: __traits(getAttributes, fun)
+  .length.iota.map!(
+    formatIndex!("@(__traits(getAttributes, %s)[%%d])".format(localSymbol))).array,
+  attributes: functionAttributes!(fun),
+  static_: __traits(isStaticFunction, fun) && isAggregate!(__traits(parent, fun)),
+  body_: ";",
+  };
+
+  return model;
+}
+
+private enum isScope(alias T) = is(T == module) || is(T == struct)
+  || is(T == class) || is(T == interface);
+
+Function refract(
+  alias Scope, string localSymbol, string name, uint index)()
+  if (isScope!Scope)
+{
+  return refract!(
+    __traits(getOverloads, Scope, name)[index],
+    `__traits(getOverloads, %s, "%s")[%d]`.format(localSymbol, name, index))
+    .setIndex(index);
+}
+
+private enum True(T...) = true;
+
+Function[] refract(alias Scope, string localSymbol, alias Predicate = True)()
+  if (isScope!Scope)
+{
+  Function[] functions;
+
+  static foreach (member; __traits(allMembers, Scope)) {
+    static foreach (index, fun; __traits(getOverloads, Scope, member)) {
+      static if (Predicate!fun) {
+        functions ~= refract!(Scope, localSymbol, member, index);
+      }
+    }
+  }
+
+  return functions;
+}
+
+///
+unittest
+{
+  pure @nogc int answer(lazy string question);
+  alias F = answer; // typically F is a template argument
+  static assert(
+    refract!(F, "F").mixture ==
+    "pure @nogc @system std.traits.ReturnType!(F) answer(lazy std.traits.Parameters!(F)[0] _0);");
+}
+
+///
+unittest
+{
+  import std.format;
+  import std.traits;
+
+  interface GrandTour
+  {
+    pure int foo() immutable;
+    @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int) const;
+    @safe shared scope void bar(scope Object);
+  }
+
+  class Mock(Interface) : Interface
+  {
+    static foreach (member; __traits(allMembers, Interface)) {
+        static foreach (fun; __traits(getOverloads, Interface, member)) {
+          mixin({
+              enum Model = refract!(fun, "fun");
+              if (is(ReturnType!fun == void)) {
+                return Model.setBody("{}").mixture;
+              } else if (Model.attributes & FunctionAttribute.ref_) {
```

Comment:
Would you have to do something similar of there was a `FunctionAttribute.static_`?

Reply:
> Would you have to do something similar of there was a `FunctionAttribute.static_`?

No. Here we test the ref-ness, not to reproduce the function, but to manufacture a legal return value.